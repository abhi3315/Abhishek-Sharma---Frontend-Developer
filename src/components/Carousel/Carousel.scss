.carousel-component {
	.carousel {
		display: inline-flex;
		overflow-x: hidden;
		scroll-snap-type: x mandatory;
		-webkit-overflow-scrolling: touch;
		scrollbar-width: none;
		-ms-overflow-style: none;
		overflow: hidden;
		height: 100%;
	}
}
