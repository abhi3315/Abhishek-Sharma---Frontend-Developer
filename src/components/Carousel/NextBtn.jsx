import PropTypes from 'prop-types';

function NextBtn({ handleClick }) {
	return (
		<button
			className='absolute top-0 bottom-0 right-0 z-[1] flex w-[15%] items-center justify-center border-0 bg-none p-0 text-center text-white opacity-50 transition-opacity duration-150 ease-[cubic-bezier(0.25,0.1,0.25,1.0)] hover:text-white hover:no-underline hover:opacity-90 hover:outline-none focus:text-white focus:no-underline focus:opacity-90 focus:outline-none motion-reduce:transition-none'
			type='button'
			aria-label='Next'
			onClick={handleClick}
		>
			<span className='inline-block h-8 w-8'>
				<svg
					xmlns='http://www.w3.org/2000/svg'
					fill='none'
					viewBox='0 0 24 24'
					strokeWidth={1.5}
					stroke='currentColor'
					className='w-6 h-6'
				>
					<path
						strokeLinecap='round'
						strokeLinejoin='round'
						d='M11.25 4.5l7.5 7.5-7.5 7.5m-6-15l7.5 7.5-7.5 7.5'
					/>
				</svg>
			</span>
			<span className='!absolute !-m-px !h-px !w-px !overflow-hidden !whitespace-nowrap !border-0 !p-0 ![clip:rect(0,0,0,0)]'>
				Next
			</span>
		</button>
	);
}

NextBtn.propTypes = {
	handleClick: PropTypes.func.isRequired,
};

export default NextBtn;
