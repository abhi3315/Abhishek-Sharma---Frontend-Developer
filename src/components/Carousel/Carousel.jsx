import { useState, createRef, useEffect } from 'react';
import PropTypes from 'prop-types';

import PrevBtn from './PrevBtn';
import NextBtn from './NextBtn';
import './Carousel.scss';

function Carousel({ images, controls, interval, pause }) {
	const [activeSlide, setActiveSlide] = useState(0);

	// Total number of images
	const totalImages = images.length;

	// Create refs for each image
	const refs = images.reduce((acc, val, index) => {
		acc[index] = createRef();
		return acc;
	}, {});

	/**
	 * Function to scroll to a specific slide
	 *
	 * @param {number} slideIndex Slide index to scroll to
	 */
	const scrollToImage = (slideIndex) => {
		setActiveSlide(slideIndex);
		refs[slideIndex].current.scrollIntoView({
			behavior: 'smooth',
			block: 'nearest',
			inline: 'start',
		});
	};

	/**
	 * Function to scroll to the next slide
	 */
	const nextImage = () => {
		if (activeSlide >= totalImages - 1) {
			scrollToImage(0);
		} else {
			scrollToImage(activeSlide + 1);
		}
	};

	/**
	 * Function to scroll to the previous slide
	 */
	const previousImage = () => {
		if (activeSlide === 0) {
			scrollToImage(totalImages - 1);
		} else {
			scrollToImage(activeSlide - 1);
		}
	};

	useEffect(() => {
		// Scroll to the active slide
		scrollToImage(activeSlide);

		if (pause) {
			return;
		}

		// Set the interval
		const intervalId = setInterval(() => {
			nextImage();
		}, interval);

		// Clear the interval on unmount
		return () => clearInterval(intervalId);
	}, [activeSlide, pause, interval]);

	return (
		<div className='flex h-full justify-center w-full items-center carousel-component'>
			<div className='relative w-full h-full'>
				<div className='carousel'>
					{controls && <PrevBtn handleClick={previousImage} />}
					{images.map((src, index) => (
						<div className='w-full flex-shrink-0' key={src} ref={refs[index]}>
							<img
								src={src}
								alt={`slide-${index + 1}`}
								className='w-full h-full object-bottom object-cover'
								loading='lazy'
							/>
						</div>
					))}
					{controls && <NextBtn handleClick={nextImage} />}
				</div>
			</div>
		</div>
	);
}

Carousel.propTypes = {
	images: PropTypes.arrayOf(PropTypes.string).isRequired,
	controls: PropTypes.bool,
	interval: PropTypes.number,
	pause: PropTypes.bool,
};

Carousel.defaultProps = {
	controls: true,
	interval: 5000,
	pause: true,
};

export default Carousel;
