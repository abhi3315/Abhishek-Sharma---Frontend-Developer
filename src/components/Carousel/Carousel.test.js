import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import Carousel from './Carousel';

describe('Carousel', () => {
	it('should render correctly', () => {
		const { container } = render(<Carousel />);
		expect(container).toMatchSnapshot();
	});

	it('should render the correct number of images', () => {
		const images = ['https://imgur.com/DaCfMsj.jpg', 'https://imgur.com/azYafd8.jpg'];
		const { getAllByRole } = render(<Carousel images={images} />);
		const renderedImages = getAllByRole('img');
		expect(renderedImages).toHaveLength(images.length);
	});
});
