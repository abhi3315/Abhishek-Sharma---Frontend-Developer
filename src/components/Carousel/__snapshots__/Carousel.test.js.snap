// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render correctly 1`] = `
<div>
  <div
    class="flex h-full justify-center w-full items-center carousel-component"
  >
    <div
      class="relative w-full h-full"
    >
      <div
        class="carousel"
      >
        <button
          aria-label="Previous"
          class="absolute top-0 bottom-0 left-0 z-[1] flex w-[15%] items-center justify-center border-0 bg-none p-0 text-center text-white opacity-50 transition-opacity duration-150 ease-[cubic-bezier(0.25,0.1,0.25,1.0)] hover:text-white hover:no-underline hover:opacity-90 hover:outline-none focus:text-white focus:no-underline focus:opacity-90 focus:outline-none motion-reduce:transition-none"
          type="button"
        >
          <span
            class="inline-block h-8 w-8"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18.75 19.5l-7.5-7.5 7.5-7.5m-6 15L5.25 12l7.5-7.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
          <span
            class="!absolute !-m-px !h-px !w-px !overflow-hidden !whitespace-nowrap !border-0 !p-0 ![clip:rect(0,0,0,0)]"
          >
            Previous
          </span>
        </button>
        <button
          aria-label="Next"
          class="absolute top-0 bottom-0 right-0 z-[1] flex w-[15%] items-center justify-center border-0 bg-none p-0 text-center text-white opacity-50 transition-opacity duration-150 ease-[cubic-bezier(0.25,0.1,0.25,1.0)] hover:text-white hover:no-underline hover:opacity-90 hover:outline-none focus:text-white focus:no-underline focus:opacity-90 focus:outline-none motion-reduce:transition-none"
          type="button"
        >
          <span
            class="inline-block h-8 w-8"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.25 4.5l7.5 7.5-7.5 7.5m-6-15l7.5 7.5-7.5 7.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
          <span
            class="!absolute !-m-px !h-px !w-px !overflow-hidden !whitespace-nowrap !border-0 !p-0 ![clip:rect(0,0,0,0)]"
          >
            Next
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
